### PR TITLE
feat(GTM-2): Add multi-cast narrator filter toggle

### PR DESCRIPTION
**Pull request by Isuru-F**
## 📖 Summary

Implements multi-cast narrator filtering toggle for audiobook discovery. Allows users to filter audiobooks that have multiple narrators/voice actors, making it easier to find multi-cast performances.

**Linear Issue:** [GTM-2](https://linear.app/sourcegraph/issue/GTM-2/add-multi-cast-narrator-support) - Add multi-cast narrator support

## ✨ Product Manager Summary

- Users can now discover multi-cast audiobooks using a toggle filter
- Toggle appears next to the search bar with clear visual feedback
- Filter works independently or combined with existing search functionality  
- Improves audiobook discovery experience for users preferring diverse voice acting
- Visual state indicators show when filter is active vs inactive

## 🛠️ Technical Notes

**Implementation Approach:** Option 2 from technical review - Separate Multi-Cast Filter Function

**Key Changes:**
- Added `isMultiCastAudiobook()` helper function for clean separation of concerns
- Extended `filteredAudiobooks` computed property with array chaining for composable filters
- Added reactive `multiCastOnly` toggle state with Vue 3.5 composition API
- Implemented custom toggle component with gradient styling matching app theme
- Updated filtering logic to handle narrator data types (string vs object)
- Enhanced no-results messaging for multi-cast specific feedback

**Architecture Benefits:**
- Reusable filter function for future extensions
- Clean separation of filtering concerns
- Maintainable and testable code structure
- Follows existing Vue 3.5/TypeScript patterns

## 📊 Feature Workflow

```mermaid
graph LR
    A[User visits audiobooks page] --> B{Multi-Cast Toggle?}
    B -->|OFF| C[Show all audiobooks]
    B -->|ON| D[Filter audiobooks with >1 narrator]
    C --> E{Search query?}
    D --> E{Search query?}
    E -->|No| F[Display results]
    E -->|Yes| G[Apply text search to results]
    G --> F[Display results]
    F --> H{Results found?}
    H -->|Yes| I[Show audiobook grid]
    H -->|No| J[Show context-aware no results message]
```

## ✅ Testing Summary

**Added 0 tests, removed 0 tests**

No new unit tests were added as per instruction to focus on visual functionality testing. All testing was performed through visual verification at localhost:5173.

## 🧪 Human Testing Instructions

1. **Base functionality:** Visit http://localhost:5173/ and verify all audiobooks display initially
2. **Toggle activation:** Click "Multi-Cast Only" toggle next to search bar
   - **Expected:** Only audiobooks with multiple narrators should display (currently "Offside" and "The Paradise Problem")
   - **Expected:** Toggle should show purple gradient when active
3. **Search combination:** With toggle enabled, search for "christina"
   - **Expected:** Only "The Paradise Problem" by Christina Lauren should display 
   - **Expected:** Combines both multi-cast filter + text search correctly
4. **Toggle deactivation:** Click toggle again to turn off
   - **Expected:** All audiobooks return to view
   - **Expected:** Toggle returns to grey inactive state
5. **No results message:** Enable toggle and search for non-existent multi-cast book
   - **Expected:** Should show "No audiobooks match your criteria" message

## 🎯 Acceptance Criteria Validation

- ✅ Multi-Cast Only toggle displays next to search bar
- ✅ When enabled, only audiobooks with >1 narrator are shown 
- ✅ Toggle state persists during search operations
- ✅ Toggle combines with text search functionality
- ✅ Toggle shows visual indication of active state (purple gradient)
- ✅ User sees context-aware feedback when no multi-cast audiobooks match criteria

### CODE REVIEW

File: client/src/views/AudiobooksView.vue [hunk 1]
```diff
@@ -5,36 +5,53 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
 
 const spotifyStore = useSpotifyStore();
 const searchQuery = ref('');
+const multiCastOnly = ref(false);
+
+// Helper function to check if audiobook has multiple narrators
+const isMultiCastAudiobook = (audiobook: typeof spotifyStore.audiobooks[0]): boolean => {
+  if (!audiobook.narrators || !Array.isArray(audiobook.narrators)) {
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:12]: **🟡 MEDIUM - ��� HIGH - Missing Unit Tests for New Feature**

The new multi-cast filtering functionality lacks unit tests. The `isMultiCastAudiobook` helper function and the updated `filteredAudiobooks` computed property need comprehensive test coverage.

**Suggestion:** Add unit tests to verify multi-cast detection logic, filtering behavior, and edge cases like missing narrator arrays.

---
*🤖 Generated by AMP Code Review Bot*

Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:12]: **🟡 MEDIUM - ��� MEDIUM - Weak Type Safety**

Using `typeof spotifyStore.audiobooks[0]` for typing is fragile and may break if the array is empty. This creates runtime risk.

**Suggestion:** Define a proper TypeScript interface for the audiobook type or import it from the store/types file.

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 2]
```diff
@@ -5,36 +5,53 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
...
+    return false;
+  }
+  return audiobook.narrators.length > 1;
+};
 
 const filteredAudiobooks = computed(() => {
-  if (!searchQuery.value.trim()) {
-    return spotifyStore.audiobooks;
+  let result = spotifyStore.audiobooks;
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:19]: **🟡 MEDIUM - ��� MEDIUM - Performance Inefficiency**

The filtering logic applies filters sequentially with separate `.filter()` calls, creating unnecessary intermediate arrays for large datasets.

**Suggestion:** Combine filters into a single `.filter()` call with combined conditions for better performance.

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 3]
```diff
@@ -67,7 +97,9 @@ onMounted(() => {
         <button @click="spotifyStore.fetchAudiobooks()">Try Again</button>
       </div>
       <div v-else>
-        <p v-if="filteredAudiobooks.length === 0" class="no-results">No audiobooks match your search.</p>
+        <p v-if="filteredAudiobooks.length === 0" class="no-results">
+          {{ multiCastOnly && !searchQuery.trim() ? 'No multi-cast audiobooks found.' : 'No audiobooks match your criteria.' }}
+        </p>
         <div v-else class="audiobook-grid">
           <AudiobookCard 
             v-for="audiobook in filteredAudiobooks"
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:105]: **🟡 MEDIUM - ��� MEDIUM - Accessibility Missing**

The custom toggle lacks proper accessibility attributes like `aria-label`, `role`, and focus management for screen readers.

**Suggestion:** Add `aria-label="Filter for multi-cast audiobooks only"` and ensure keyboard navigation works properly.

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 4]
```diff
@@ -67,7 +97,9 @@ onMounted(() => {
         <button @click="spotifyStore.fetchAudiobooks()">Try Again</button>
       </div>
       <div v-else>
-        <p v-if="filteredAudiobooks.length === 0" class="no-results">No audiobooks match your search.</p>
+        <p v-if="filteredAudiobooks.length === 0" class="no-results">
+          {{ multiCastOnly && !searchQuery.trim() ? 'No multi-cast audiobooks found.' : 'No audiobooks match your criteria.' }}
+        </p>
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:102]: **🟡 MEDIUM - ��� LOW - Code Complexity**

The conditional message logic in the template is becoming complex and could be harder to maintain.

**Suggestion:** Extract the message logic into a computed property for better readability and testability.

---
*🤖 Generated by AMP Code Review Bot*

**Review by amp-agent-isuru — CHANGES_REQUESTED** [review]
## 🤖 AMP Code Review

Found **5** issues that need attention. See inline comments for details.

**Issues Summary:**
1. 🟡 MEDIUM - ��� HIGH - Missing Unit Tests for New Feature (client/src/views/AudiobooksView.vue:12)
2. 🟡 MEDIUM - ��� MEDIUM - Weak Type Safety (client/src/views/AudiobooksView.vue:12)
3. 🟡 MEDIUM - ��� MEDIUM - Performance Inefficiency (client/src/views/AudiobooksView.vue:19)
4. 🟡 MEDIUM - ��� MEDIUM - Accessibility Missing (client/src/views/AudiobooksView.vue:105)
5. 🟡 MEDIUM - ��� LOW - Code Complexity (client/src/views/AudiobooksView.vue:102)

**Next Steps:**
- Address each issue marked in the code
- Resolve conversations when issues are fixed
- Re-request review when all issues are addressed

---
*Review completed on 2025-08-15*
